ide/gsf.testrunner.ui/src/org/netbeans/modules/gsf/testrunner/ui/TestMethodFinderImpl.java

```diff
@@ -0,0 +1,166 @@
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.BiConsumer;
+import java.util.logging.Logger;
+import org.netbeans.api.editor.mimelookup.MimeLookup;
+import org.netbeans.api.editor.mimelookup.MimeRegistration;
+import org.netbeans.modules.gsf.testrunner.ui.api.TestMethodController;
+import org.netbeans.modules.gsf.testrunner.ui.spi.ComputeTestMethods;
+import org.netbeans.modules.parsing.api.Snapshot;
+import org.netbeans.modules.parsing.spi.Parser;
+import org.netbeans.modules.parsing.spi.indexing.Context;
+import org.netbeans.modules.parsing.spi.indexing.EmbeddingIndexer;
+import org.netbeans.modules.parsing.spi.indexing.EmbeddingIndexerFactory;
+import org.netbeans.modules.parsing.spi.indexing.Indexable;
+import org.openide.filesystems.FileObject;
+import org.openide.filesystems.FileUtil;
+import org.openide.util.Exceptions;
+import org.openide.util.WeakSet;
+
+/**
+ *
+ * @author Dusan Balek
+ */
+public final class TestMethodFinderImpl extends EmbeddingIndexer {
+
+    public static final String NAME = "tests"; // NOI18N
+    public static final int VERSION = 1;
+    public static final TestMethodFinderImpl INSTANCE = new TestMethodFinderImpl();
+
+    private final WeakSet<BiConsumer<FileObject, Collection<TestMethodController.TestMethod>>> listeners = new WeakSet<>();
+
+    @Override
+    protected void index(Indexable indexable, Parser.Result parserResult, Context context) {
+        List<TestMethodController.TestMethod> testMethods = new ArrayList<>();
+        for (ComputeTestMethods ctm : MimeLookup.getLookup(indexable.getMimeType()).lookupAll(ComputeTestMethods.class)) {
+            testMethods.addAll(ctm.computeTestMethods(parserResult, new AtomicBoolean()));
+        }
+        FileObject fo = parserResult.getSnapshot().getSource().getFileObject();
+        store(context.getIndexFolder(), indexable.getURL(), indexable.getRelativePath(), testMethods);
+        if (!context.isAllFilesIndexing()) {
+            synchronized (listeners) {
+                for (BiConsumer<FileObject, Collection<TestMethodController.TestMethod>> listener : listeners) {
+                    listener.accept(fo, testMethods);
+                }
+            }
+        }
+    }
+
+    public void addListener(BiConsumer<FileObject, Collection<TestMethodController.TestMethod>> listener) {
+        synchronized(listeners) {
+            listeners.putIfAbsent(listener);
+            Logger.getLogger(TestMethodFinderImpl.class.getName()).info("Listener added: " + listener);
+        }
+    }
+
+    private void store(FileObject indexFolder, URL url, String resourceName, List<TestMethodController.TestMethod> methods) {
+        File cacheRoot = FileUtil.toFile(indexFolder);
+        File output = new File(cacheRoot, resourceName + ".tests"); //NOI18N
+        if (methods.isEmpty()) {
+            if (output.exists()) {
+                output.delete();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of deleting a file on this line should be checked and failure should be logged. java.io.File.delete() will simply return false if it fails to delete the file, for example, when the file does not exist. It will not throw unless there is a security exception.